### ISSUE TITLE
Increase verbosity of 'you are not authorized' message

### ISSUE THREAD

**Issue by Gryllida** [atheme/atheme#573]
STR:

> /msg chanserv set #chan guard on

ACTUAL RESULT:

>Chanserv<You are not authorized to perform this command.

EXPECTED RESULT:

>Chanserv<You are not authorized to perform this command (requires flag +s; you have none)

(Or whatever I have)

Or

>Chanserv<You are not authorized to perform this command (requires flag +s; you have +Ffo)
>Chanserv<(Note: as you have +F, you can add +s to your ACL entry using the FLAGS command.)

**Comment by Zarthus** [comment]
this sounds rather confusing for average irc users, and also very easy to date or get wrong. Perhaps this should only be used if some sort of user-setting "expert user" is enabled (that networks can enable or disable by default for new registrations)?

**Comment by thesamesam** [comment]
@Zarthus: I can agree with your comments for the second message but I don't know if a small hint as to which flags would be needed would confuse anyone. The user is already trying to set a flag on a channel so making the flags less opaque by e.g. clarifying what the user needs to ask for isn't going to hurt.

The second example is a bit more spam-like and would be excessive but I think a small hint would only add clarity, and not introduce any harm.

**Comment by Zarthus** [comment]
Unless I am misunderstanding the suggestion, it would apply to all unauthorized messages, and not just a few of them. They are not setting a flag on the channel with GUARD, the error could read you need +f for FLAGS, you need +r for ChanServ/KICK, and +o for ChanServ/OP -- and so on.

So no, the user is not trying to set a flag on a channel, and when I am getting opped and the services tell me I need the flag +o, I'll just be confused as a regular user.

To tag on to my last statement, imagining someone very new to IRC would think that it's ridiculous they need +o to get op, as once you have +o you already are op. ("You need flag +o to use ChanServ/OP")

**Comment by Gryllida** [comment]
Opt-in would be ok?

Another option, instead or complementary, is to only give this information to people who already are on the access list of the target channel.

**Comment by grawity** [comment]
@Zarthus I really don't understand your "the user is not trying to set a flag on a channel" argument... Are you trying to say that error messages cannot mention any prerequisites for doing an action?

... honestly, that post is literally below zero on the "makes sense" scale. Mind rephrasing?

---

(IMHO, people who are "very new to IRC" generally won't be using such ChanServ features yet; and people who use ChanServ commands will sooner or later have to learn the difference between flags and modes anyway. At minimum, having something copypastable to \#irchelp would certainly speed things up.)

**Comment by Zarthus** [comment]
That's not what I meant at all, I'm saying that a GUARD is not setting flags on a channel (only ChanServ/flags sets flags on a channel)

The flag for ops (+o) is the same as the channel mode (+o). When you as an user see the services suggest you need +o it would be rather nonsensical if you don't understand or are aware of the flags system. You'd think you need the channel mode (channel op) to get channel op.

Perhaps if we instead mention a permission system with flag names (e.g. chan:auspex(+A) ) instead of mentioning +A directly it might become clearer to the user, but I don't know. it's why I'm suggesting the "advanced user" mode.

I'm absolutely in favour of errors mentioning the prerequisites, I just think the way it was being phrased in the issue description was going to lead to confusion for new users.

**Comment by thesamesam** [comment]
Sorry. To clarify, when I meant 'setting flags', I meant this in the non-atheme sense: making a setting change to a channel. It is expected that they may not have privileges or need to know which privileges they are missing.

**Comment by juggs** [comment]
I agree with the phrasing being potentially confusing. On the other hand I'm not convinced presenting a 'new user' with a message relating to 
e.g. chan:auspex(+A) really helps matters either unless the text is verbose or links out to some reference materials - and even then 
understanding the layering of modes and permissions is not something said 'new user' is likely to want to entertain to just get something done stat.

Always the problem with documentation around ircds and services is that by the time anyone is in a vague position of knowledge to begin 
documenting, they by necessity already know too much to document it effectively from a 'new user' perspective as that perspective is lost in the 
sands of time that is our fallible memories.

It's why we end up with such nuggets as:

https://github.com/atheme/atheme/wiki/ChanServ%3ASET%3AGUARD
ChanServ:SET:GUARD
SET GUARD allows you to have ChanServ join your channel.

True enough of course. But 'new user' is left wondering "Why would I want ChanServ to join my channel? And what has ChanServ (whatever that 
maybe) got to do with guarding my channel - don't get it :/".

As always, I come with no answers, probably just restating the obvious.

G

**Comment by ilbelkyr** [comment]
fwiw I mostly like what we currently do for commands that require soper access:

> You are not authorized to perform this operation.

for regular, non-privileged users;

> You do not have the user:auspex privilege.

for people with some sort of special access by virtue of being a services or IRC operator.

Expanding on that idea, we could have

> You do not have any special access to **#channel**.

vs

> You do not have the **+s** flag for **#channel**.

(or whatever wording we come up with, that's just an example).

As @Zarthus pointed out above, this would need to be kept up to date manually for now. I wonder if it might not be more appropriate to have a more declarative method of handling channel flags (similar to how commands can require a flag in the `command_t` itself). There might still be some special cases a simple approach can't handle, but we have that with soper privileges as well already in some places.
